File: Assemblies/Protean.Tools/Http/Http.WebRequest.vb
Option Strict Off
Option Explicit On

Imports System
Imports System.IO
Imports System.Net
Imports System.Web.HttpUtility
Imports System.Text
Imports System.Text.Encoding



Namespace Http
    ''' <summary>
    ''' Simplified WebRequest class.
    ''' </summary>
    ''' <remarks>TODO: Needs a GET method</remarks>
    Public Class WebRequest

#Region "Declarations"

        Private _moduleName As String = "Protean.Tools.WebRequest"

        Private _contentType As String = "text/html"
        Private _contentLength As Long = 0
        Private _userAgent As String = ""
        Private _requestEncoding As New ASCIIEncoding
        Private _requestBody As String
        Private _requestBytes As Byte()
        Private _method As String = "POST"
        Private _includeResponse As Boolean = True

#End Region

#Region "Events"

        Public Event OnError(ByVal sender As Object, ByVal e As Protean.Tools.Errors.ErrorEventArgs)

#End Region

#Region "Constructor"

        Public Sub New()

        End Sub


        Public Sub New(ByVal contentType As String)
            Me.ContentType() = contentType
        End Sub

        Public Sub New(ByVal contentType As String, ByVal method As String)
            Me.ContentType() = contentType
            Me.Method() = method
        End Sub

        Public Sub New(ByVal contentType As String, ByVal method As String, ByVal userAgent As String)
            Me.ContentType() = contentType
            Me.Method() = method
            Me.UserAgent() = userAgent
        End Sub

#End Region

#Region "Private Properties"

#End Region

#Region "Public Properties"

        Public Property ContentType() As String
            Get
                Return _contentType
            End Get
            Set(ByVal value As String)
                _contentType = value
            End Set
        End Property

        Public Property Method() As String
            Get
                Return _method
            End Get
            Set(ByVal value As String)
                _method = value
            End Set
        End Property

        Public Property UserAgent() As String
            Get
                Return _userAgent
            End Get
            Set(ByVal value As String)
                _userAgent = value
            End Set
        End Property

        Public Property IncludeResponse() As Boolean
            Get
                Return _includeResponse
            End Get
            Set(ByVal value As Boolean)
                _includeResponse = value
            End Set
        End Property


        Public Property RequestBody() As String
            Get
                Return _requestBody
            End Get
            Set(ByVal value As String)
                _requestBody = value
                _requestBytes = _requestEncoding.GetBytes(_requestBody)
                _contentLength = _requestBytes.Length
            End Set
        End Property

        Public ReadOnly Property ContentLength() As Integer
            Get
                Return _contentLength
            End Get
        End Property
#End Region

#Region "Private Members"

#End Region

#Region "Public Members"

        ''' <summary>
        ''' Instigate a synchronous request and response
        ''' </summary>
        ''' <param name="url">The URI to send the request to</param>
        ''' <param name="request">The request string to send</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function Send(ByVal url As String, Optional ByVal request As String = "") As String

            Try
                If Uri.IsWellFormedUriString(url, UriKind.RelativeOrAbsolute) Then
                    Return Send(New Uri(url), request)
                Else
                    Return ""
                End If
            Catch ex As Exception
                RaiseEvent OnError(Me, New Protean.Tools.Errors.ErrorEventArgs(_moduleName, "Send(String)", ex, ""))
                Return ""
            End Try

        End Function

        ''' <summary>
        ''' Instigate a synchronous request and response
        ''' </summary>
        ''' <param name="url">The URI to send the request to</param>
        ''' <param name="request">The request string to send</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function Send(ByVal url As Uri, Optional ByVal request As String = "") As String
            Try

                ' Create the response object
                Dim newRequest As HttpWebRequest = Nothing
                Dim newResponse As HttpWebResponse = Nothing
                Dim requestStream As Stream = Nothing
                Dim responseStream As StreamReader = Nothing
                Dim responseOutput As String = ""

                If Not (String.IsNullOrEmpty(request)) Then RequestBody = request

                If Not (Me.ContentLength = 0 And Me.Method = "POST") Then
                    ' Create the request object
                    newRequest = HttpWebRequest.Create(url)

                    ' Set the standard headers
                    newRequest.ContentType = Me.ContentType
                    newRequest.ContentLength = Me.ContentLength
                    newRequest.Method = Me.Method
                    'newRequest.MaximumAutomaticRedirections = 4
                    'newRequest.MaximumResponseHeadersLength = 4

                    ' Set the optional headers
                    If Not (String.IsNullOrEmpty(Me.UserAgent)) Then newRequest.UserAgent = Me.UserAgent

                    ' POST the request body
                    If Method = "POST" Then
                        requestStream = newRequest.GetRequestStream
                        requestStream.Write(Me._requestBytes, 0, Me.ContentLength)
                        requestStream.Close()
                    End If


                    If IncludeResponse() Then
                        ' Get the response
                        newResponse = newRequest.GetResponse()
                        requestStream = newResponse.GetResponseStream()
                        responseStream = New StreamReader(requestStream, System.Text.Encoding.UTF8)
                        responseOutput = responseStream.ReadToEnd

                        requestStream.Close()
                        responseStream.Close()
                        newResponse.Close()
                    End If

                End If

                Return responseOutput

            Catch ex As Exception
                RaiseEvent OnError(Me, New Protean.Tools.Errors.ErrorEventArgs(_moduleName, "Send(Uri)", ex, ""))
                Return ""
            End Try
        End Function

#End Region

    End Class

End Namespace

File: Assemblies/Protean.Tools/Http/Http.WebRequest.cs

using System;
using System.IO;
using System.Net;
using System.Text;
using static System.Text.Encoding;

namespace Protean.Tools.Http
{
    /// <summary>
    /// Simplified WebRequest class.
    /// </summary>
    /// <remarks>TODO: Needs a GET method</remarks>
    public partial class WebRequest
    {

        #region Declarations

        private string _moduleName = "Protean.Tools.WebRequest";

        private string _contentType = "text/html";
        private long _contentLength = 0L;
        private string _userAgent = "";
        private ASCIIEncoding _requestEncoding = new ASCIIEncoding();
        private string _requestBody;
        private byte[] _requestBytes;
        private string _method = "POST";
        private bool _includeResponse = true;

        #endregion

        #region Events

        public event OnErrorEventHandler OnError;

        public delegate void OnErrorEventHandler(object sender, Protean.Tools.Errors.ErrorEventArgs e);

        #endregion

        #region Constructor

        public WebRequest()
        {

        }


        public WebRequest(string contentType)
        {
            ContentType = contentType;
        }

        public WebRequest(string contentType, string method)
        {
            ContentType = contentType;
            Method = method;
        }

        public WebRequest(string contentType, string method, string userAgent)
        {
            ContentType = contentType;
            Method = method;
            UserAgent = userAgent;
        }

        #endregion

        #region Private Properties

        #endregion

        #region Public Properties

        public string ContentType
        {
            get
            {
                return _contentType;
            }
            set
            {
                _contentType = value;
            }
        }

        public string Method
        {
            get
            {
                return _method;
            }
            set
            {
                _method = value;
            }
        }

        public string UserAgent
        {
            get
            {
                return _userAgent;
            }
            set
            {
                _userAgent = value;
            }
        }

        public bool IncludeResponse
        {
            get
            {
                return _includeResponse;
            }
            set
            {
                _includeResponse = value;
            }
        }


        public string RequestBody
        {
            get
            {
                return _requestBody;
            }
            set
            {
                _requestBody = value;
                _requestBytes = _requestEncoding.GetBytes(_requestBody);
                _contentLength = _requestBytes.Length;
            }
        }

        public int ContentLength
        {
            get
            {
                return (int)_contentLength;
            }
        }
        #endregion

        #region Private Members

        #endregion

        #region Public Members

        /// <summary>
        /// Instigate a synchronous request and response
        /// </summary>
        /// <param name="url">The URI to send the request to</param>
        /// <param name="request">The request string to send</param>
        /// <returns></returns>
        /// <remarks></remarks>
        public string Send(string url, string request = "")
        {

            try
            {
                if (Uri.IsWellFormedUriString(url, UriKind.RelativeOrAbsolute))
                {
                    return Send(new Uri(url), request);
                }
                else
                {
                    return "";
                }
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(_moduleName, "Send(String)", ex, ""));
                return "";
            }

        }

        /// <summary>
        /// Instigate a synchronous request and response
        /// </summary>
        /// <param name="url">The URI to send the request to</param>
        /// <param name="request">The request string to send</param>
        /// <returns></returns>
        /// <remarks></remarks>
        public string Send(Uri url, string request = "")
        {
            try
            {

                // Create the response object
                HttpWebRequest newRequest = default;
                HttpWebResponse newResponse = default;
                Stream requestStream = null;
                StreamReader responseStream = null;
                string responseOutput = "";

                if (!string.IsNullOrEmpty(request))
                    RequestBody = request;

                if (!(ContentLength == 0 & Method == "POST"))
                {
                    // Create the request object
                    newRequest = (HttpWebRequest)HttpWebRequest.Create(url);

                    // Set the standard headers
                    newRequest.ContentType = ContentType;
                    newRequest.ContentLength = ContentLength;
                    newRequest.Method = Method;
                    // newRequest.MaximumAutomaticRedirections = 4
                    // newRequest.MaximumResponseHeadersLength = 4

                    // Set the optional headers
                    if (!string.IsNullOrEmpty(UserAgent))
                        newRequest.Headers["User-Agent"] = UserAgent;

                    // POST the request body
                    if (Method == "POST")
                    {
                        requestStream = newRequest.GetRequestStream();
                        requestStream.Write(_requestBytes, 0, ContentLength);
                        requestStream.Close();
                    }


                    if (IncludeResponse)
                    {
                        // Get the response
                        newResponse = (HttpWebResponse)newRequest.GetResponse();
                        requestStream = newResponse.GetResponseStream();
                        responseStream = new StreamReader(requestStream, UTF8);
                        responseOutput = responseStream.ReadToEnd();

                        requestStream.Close();
                        responseStream.Close();
                        newResponse.Close();
                    }

                }

                return responseOutput;
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(_moduleName, "Send(Uri)", ex, ""));
                return "";
            }
        }

        #endregion

    }

}
